src/hotspot/share/ci/ciInstanceKlass.hpp

```diff
@@ -152,16 +152,15 @@ class ciInstanceKlass : public ciKlass {
     return _has_finalizer; }
   bool                   has_subklass()   {
     assert(is_loaded(), "must be loaded");
-    if (_has_subklass == subklass_unknown ||
-        (_is_shared && _has_subklass == subklass_false)) {
-      if (flags().is_final()) {
-        return false;
-      } else {
-        return compute_shared_has_subklass();
-      }
+    if (_has_subklass == subklass_true) {
```

Comment:
A comment explaining why we only cache `subklass_true` would be good. Also, do we still need a tri-state including `subklass_unknown`? It's now only used for initialization but not queried anywhere.

Reply:
Comment added.

Regarding `subclass_false` and `subklass_unknown`, keeping tri-state enum looks benign. Also, it makes sense to keep `subklass_unknown` to be able to distinguish between not yet initialized and already initialized cases (at least, during debugging).

Reply:
Okay, looks good.